### PR TITLE
feat: add support to consumer_groups

### PR DESCRIPTION
**Pull request by GGabriele**
https://docs.konghq.com/gateway/3.0.x/admin-api/consumer-groups/reference/

Needs some go-kong changes first: https://github.com/Kong/go-kong/pull/221

### CODE REVIEW

**Review by aboudreault — DISMISSED** [review]
Great job @GGabriele !! I looked at all files and I don't see any blocker. There is a lot in this PR and it's not obvious to catch possible bugs while reviewing. However, I tested this deck PR with Konnect with multiple different cases (sync/dump/diff etc.) and things seem to work well. I am continuing some more testing but I'd say +1.

File: dump/dump.go [hunk 1]
```diff
@@ -501,6 +525,71 @@ func GetAllUpstreams(ctx context.Context,
...
+		return nil, err
+	}
+	for _, cg := range cgs {
+		r, err := konnect.GetConsumerGroupObject(ctx, client, cg.ID)
+		if err != nil {
+			return nil, err
+		}
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply dump/dump.go:548]: Nope, it was a leftover, good catch!

File: dump/dump.go [hunk 2]
```diff
@@ -501,6 +525,71 @@ func GetAllUpstreams(ctx context.Context,
 	return upstreams, nil
 }
 
+// GetAllConsumerGroups queries Kong for all the ConsumerGroups using client.
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply dump/dump.go:528]: Indeed :D

File: dump/dump.go [hunk 3]
```diff
@@ -235,6 +236,29 @@ func getProxyConfiguration(ctx context.Context, group *errgroup.Group,
...
+		if config.KonnectRuntimeGroup != "" {
+			consumerGroups, err = GetAllKonnectConsumerGroups(ctx, client, config.SelectorTags)
+			if kong.IsNotFoundErr(err) {
+				return nil
+			}
+		} else {
+			consumerGroups, err = GetAllConsumerGroups(ctx, client, config.SelectorTags)
+			if kong.IsNotFoundErr(err) {
+				return nil
+			}
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply dump/dump.go:254]: Sure!

**Review by aboudreault — APPROVED** [review]

**Comment by sscarpanti70** [comment]
Hello @GGabriele , about consumer-groups support, with 1.17.0 deck release I cannot dump or sync from free kong anymore, while with 1.16.0 was possible. The error given was: Error: reading configuration from Kong: consumer_groups: HTTP status 403 (message: "Enterprise license missing or expired")

**Comment by GGabriele** [comment]
Hi @sscarpanti70 , what Kong version are you using?

**Comment by sscarpanti70** [comment]
3.1.1.1, without any license so far

Il giorno gio 12 gen 2023 alle ore 16:02 Gabriele ***@***.***>
ha scritto:

**Comment by sscarpanti70** [comment]
I can't understand if since 1.17.0 deck a license is needed to sync with
kong or I have to add a specific user to fix connection

Il giorno gio 12 gen 2023 alle ore 16:08 Stefano Scarpanti <
***@***.***> ha scritto:

**Comment by GGabriele** [comment]
I see the problem, you're using an enterprise version of Kong without a license and decK is not checking for this case.

I will push a fix and do path release. Thanks for raising this!

**Comment by sscarpanti70** [comment]
Hello Gabriele,
I just let you know about a workaround we found with 1.17.1, simply
specifying workspace

deck -w default sync -s setup_3x.yaml --select-tag inso-single

It works perfectly. Also dump.

Il giorno gio 12 gen 2023 alle ore 21:29 Gabriele ***@***.***>
ha scritto:

**Comment by GGabriele** [comment]
Thanks for letting me know!

In the mean time, I just opened this: https://github.com/Kong/deck/pull/821

**Pull request merged**